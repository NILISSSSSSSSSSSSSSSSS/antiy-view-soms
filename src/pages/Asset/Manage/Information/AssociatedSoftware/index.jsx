import React, { PureComponent } from 'react'
import { connect } from 'dva'
import { Search, CommonModal } from '@c'
import { Form, Table, message, Select, Input } from 'antd'
import PropTypes from 'prop-types'
import { debounce, throttle } from 'lodash'
import { TooltipFn } from '@u/common'
import api from '@/services/api'

const FormItem = Form.Item
const { Option } = Select
@Form.create()
@connect(({ asset }) => ({
  installableList: asset.installableList,
  usersByRoleCodeAndAreaIdList: asset.usersByRoleCodeAndAreaIdList
}))
class AssociatedSoftware extends PureComponent{
   state = {
     PrefixCls: 'AssociatedSoftware',
     //查询开始
     currentPage: 1,
     pageSize: 10,
     selectedRowKeys: [],
     supplier: null,
     productName: null
     //查询结束
   }

    static propTypes = {
      soVisible: PropTypes.bool,
      associds: PropTypes.array,
      associatedSoftware: PropTypes.func,
      installableList: PropTypes.object,
      usersByRoleCodeAndAreaIdList: PropTypes.array
    }

    onSelectChange = (selectedRowKeys, selectedRows) => {
      this.setState({ otherBusinessIds: selectedRowKeys, selectedRowKeys })
    };

    //提交
    onSubmit = (values) => {
      this.setState({
        supplier: values.supplier,
        productName: values.productName,
        currentPage: 1,
        pageSize: 10
      }, this.installableList)
    }

    //重置
    onReset = ()=>{
      this.setState({
        supplier: null,
        productName: null,
        currentPage: 1,
        pageSize: 10
      }, this.installableList)
    }

    //分页
    pageChange=(currentPage, pageSize)=>{
      this.setState({
        currentPage,
        pageSize
        // selectedRowKeys: []
      }, this.installableList)
    }

    getRecordsArrIds=(stringId)=>{
      let assetId = []
      this.props.associds.forEach(v=>assetId.push(v[stringId]))
      return assetId
    }

    //确定关联组件
    onConfirm = throttle(()=>{
      this.props.form.validateFields((err, values) => {
        if(err) return void(0)
        this.onConfirmCB(values)
      })
    }, 5000, { trailing: false })

    /***************副作用开始************** */

    onConfirmCB=values=>{
      const {
        associatedSoftware,
        usersByRoleCodeAndAreaIdList,
        compEffectCB
      } = this.props
      let params = {
        assetId: this.getRecordsArrIds('stringId'),
        softId: this.state.selectedRowKeys
      }
      if(params.softId.length === 0){
        message.info('请选择关联软件!')
        return void(0)
      }
      if(values.baselineConfigUserId === ' '){
        values.baselineConfigUserId = usersByRoleCodeAndAreaIdList.map(item=>item.stringId).join(',')
      }
      params.manualStartActivityRequest = {
        formData: {
          'baselineConfigUserId': values.baselineConfigUserId,
          'memo': values.memo
        }
      }
      api.batchRelation(params).then(res => {
        if(res && res.head && res.head.code === '200'){
          message.success('操作成功！')
          associatedSoftware(false)
          compEffectCB()
        }
      })
    }

    installableList=debounce(async ()=>{
      const state = this.state
      const payload = {
        baselineTemplateId: this.props.associds[0].baselineTemplateId,
        supplier: state.supplier,
        productName: state.productName,
        currentPage: state.currentPage,
        pageSize: state.pageSize
      }
      await this.props.dispatch({ type: 'asset/installableList', payload: payload })
    }, 300)

    //操作人员
    getUsersByRoleCodeAndAreaId=()=>{
      this.props.dispatch({ type: 'asset/getUsersByRoleCodeAndAreaId', payload: {
        flowNodeTag: 'config_base', flowId: 4, areaId: this.getRecordsArrIds('areaId') }
      })
    }

    render (){
      const { selectedRowKeys, currentPage, pageSize } = this.state
      const { soVisible, associatedSoftware, installableList, form: { getFieldDecorator }, usersByRoleCodeAndAreaIdList } = this.props
      const defaultFields = [
        { type: 'input', label: '厂商', placeholder: '请输入厂商', key: 'supplier', allowClear: true, maxLength: 30 },
        { type: 'input', label: '名称', placeholder: '请输入名称', key: 'productName', allowClear: true, maxLength: 30 }
      ]

      const columns = [
        {
          title: '厂商',
          dataIndex: 'supplier',
          render: text => TooltipFn (text)
        },
        {
          title: '名称',
          dataIndex: 'productName',
          render: text => TooltipFn (text)
        },
        {
          title: '版本',
          dataIndex: 'version',
          render: text => TooltipFn (text)
        },
        {
          title: '软件平台',
          dataIndex: 'softPlatform',
          render: text => TooltipFn (text)
        }
      ]

      const rowSelection = {
        selectedRowKeys,
        onChange: this.onSelectChange
      }

      const formLayout = {
        labelCol: {
          span: 8
        },
        wrapperCol: {
          span: 16
        }
      }

      return(
        <CommonModal
          title="关联软件"
          type="search"
          visible={soVisible}
          width={1200}
          oktext='保存'
          onConfirm={this.onConfirm}
          onClose={()=>associatedSoftware(false)}>
          <div className="table-wrap">
            <Search defaultFields={defaultFields} onSubmit={this.onSubmit} onReset={this.onReset}/>
            <Table
              rowKey='softwareId'
              columns={columns}
              rowSelection={rowSelection}
              dataSource={installableList.items || []}
              pagination={{
                current: currentPage,
                pageSize: pageSize,
                onShowSizeChange: this.pageChange,
                onChange: this.pageChange,
                showQuickJumper: true,
                showSizeChanger: installableList.totalRecords > 10,
                pageSizeOptions: ['10', '20', '30', '40'],
                showTotal: () => `共 ${installableList.totalRecords || 0} 条数据`,
                total: installableList.totalRecords || 0
              }}
            />
            <Form className="filter-form new-flex-layout" layout="inline" onSubmit={this.handleSubmit}>
              <FormItem {...formLayout} label='下一步骤'>
                {getFieldDecorator('666')(
                  <span>基准配置</span>
                ) }
              </FormItem>
              <FormItem {...formLayout} label='下一步执行人'>
                {getFieldDecorator('baselineConfigUserId', {
                  initialValue: ' ',
                  rules: [{ required: true, message: '请指派人员！' }]
                })(
                  <Select showSearch
                    allowClear
                    getPopupContainer={triggerNode => triggerNode.parentNode}
                    optionFilterProp="children">
                    {usersByRoleCodeAndAreaIdList.length ? <Option value=' '>全部</Option> : ''}
                    {usersByRoleCodeAndAreaIdList.map(
                      (now, index)=>(<Option value = {now.stringId} key={index + 'baselineConfigUserId'}>{now.name}</Option>))
                    }
                  </Select>
                ) }
              </FormItem>
              <FormItem {...formLayout} label='备注'>
                {getFieldDecorator('memo', {
                  rules: [
                    { message: '最多300个字符！', max: 300 }
                  ]
                })(
                  <Input autoComplete='off' className="filter-form-item" placeholder="请输入" />
                ) }
              </FormItem>
            </Form>
          </div>
        </CommonModal>
      )
    }

    async componentDidMount (){
      await Promise.all([
        this.installableList(),
        this.getUsersByRoleCodeAndAreaId()
      ])
    }

    componentWillUnmount (){
      this.props.dispatch({ type: 'asset/save', payload: { installableList: {} } })
    }
}

export default AssociatedSoftware
