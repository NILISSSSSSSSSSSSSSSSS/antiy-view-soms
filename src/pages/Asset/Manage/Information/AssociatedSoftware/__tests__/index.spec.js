import React from 'react'
import { shallow, mount } from 'enzyme'
import  AssociatedSoftware from '../index'
import PropTypes from 'prop-types'

// import debounce from 'lodash/debounce'
// import * as sinon from 'sinon'

//所有注释勿删，将来有用
jest.mock('lodash', () => ({
  debounce: jest.fn((fn => fn))
}))

const batchRelation = (params)=> {
  return Promise.resolve({
    head: { code: '200' },
    body: { }
  })
}
const api = {
  batchRelation
}

jest.mock('../../../../../../services/api')
// jest.mock('@/u/common.js')
// jest.mock('@/services/api.js')
// jest.mock('@/c/Search')
// jest.mock('@/c/CommonModal')

const props = {
  // form: {
  //   getFieldDecorator: jest.fn( opts => c => c )
  // },

  soVisible: false,
  associds: [{ baselineTemplateId: 1 }],
  associatedSoftware: jest.fn()
}

const  selectes = {
  supplier: 666,
  productName: 666,
  version: 666,
  softPlatform: 666
}

const store = {
  subscribe: () => ({
  }),
  dispatch: () =>({}),
  getState: () => ({
    asset: {
      installableList: {
        items: [
          selectes
        ],
        totalRecords: 10
      },
      usersByRoleCodeAndAreaIdList: [
        {
          stringId: '1',
          name: 'duuliy'
        }
      ]
    }
  })
}
const options = {
  context: { store },
  childContextTypes: { store: PropTypes.object.isRequired }
}

describe('<AssociatedSoftware />', () => {

  const wrapper = shallow(<AssociatedSoftware {...props}/>).dive(options).dive(options)
  const instance = wrapper.instance()
  const expectStates = (vals, valCBs)=>{
    vals.forEach((v, i)=>expect(instance.state[v]).toBe(valCBs[i]))
  }
  // console.log(instance)
  it('render <AssociatedSoftware /> component', () => {
    expect(wrapper.find('.table-wrap')).toHaveLength(1)
  })
  it('testFunc onSelectChange', () => {
    instance.onSelectChange(selectes, selectes)
    const vals = ['otherBusinessIds', 'selectedRowKeys']
    const valCBs = [selectes, selectes]
    expectStates(vals, valCBs)
  })

  it('testFunc onSubmit', () => {
    instance.onSubmit(selectes)
    const vals = ['supplier', 'productName', 'currentPage', 'pageSize']
    const valCBs = [666, 666, 1, 10]
    expectStates(vals, valCBs)
  })

  it('testFunc onReset', () => {
    instance.onReset()
    const vals = ['supplier', 'productName', 'currentPage', 'pageSize']
    const valCBs = [null, null, 1, 10]
    expectStates(vals, valCBs)
  })

  it('testFunc pageChange', () => {
    const currentPage = 2
    const pageSize = 20
    instance.pageChange(currentPage, pageSize)
    const vals = ['currentPage', 'pageSize']
    const valCBs = [2, 20]
    expectStates(vals, valCBs)
  })

  it('testFunc onConfirm', () => {
    const spyFunc = jest.spyOn(instance, 'onConfirm')
    spyFunc()
    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc onConfirmCB', () => {
    const spyFunc = jest.spyOn(instance, 'onConfirmCB')
    const valuesCom = {
      'assetId': ['37g570oTx1JL7FNSqNzaMw=='],
      'softId': ['614120741717147648'],
      'memo': '666'
    }
    const values1 = {
      ...valuesCom,
      'baselineConfigUserId': '2reo+l2cXUDmupuhJ65GTg=='
    }
    const values2 = {
      ...valuesCom,
      'baselineConfigUserId': 'all'
    }
    spyFunc(values1)
    spyFunc(values2)

    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc installableList and debounce', async () => {
    // let clock
    // clock = sinon.useFakeTimers()
    // const spyFunc = jest.spyOn(instance, 'installableList')
    // const debounced = debounce(spyFunc, 1000)
    // debounced()
    // expect(spyFunc).toHaveBeenCalledTimes(0)
    // for(let i = 0;i < 10;i++){
    //   clock.tick(500)
    //   debounced()
    // }
    // expect(spyFunc).toHaveBeenCalledTimes(0)
    // clock.tick(1000)
    // expect(spyFunc).toHaveBeenCalledTimes(1)
    // // console.log(instance.props.installableList)
    // await console.log(instance.props.dispatch({ type: 'asset/installableList', payload: {
    //   baselineTemplateId: props.associds[0].baselineTemplateId,
    //   supplier: 6666,
    //   productName: 666,
    //   currentPage: 1,
    //   pageSize: 10
    // } }))
    // clock.restore()
    // console.log(instance.props.installableList)

  })

  it('shoulder render a AssociatedSoftware',  ()=> {
    const app = wrapper.debug()
    expect(app).toMatchSnapshot()
  })
})