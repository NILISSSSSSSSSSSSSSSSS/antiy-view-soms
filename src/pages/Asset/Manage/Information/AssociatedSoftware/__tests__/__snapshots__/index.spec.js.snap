// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssociatedSoftware /> shoulder render a AssociatedSoftware 1`] = `
"<Connect(Form(commonModal)) title=\\"关联软件\\" type=\\"search\\" visible={false} width={1200} oktext=\\"保存\\" onConfirm={[Function]} onClose={[Function: onClose]}>
  <div className=\\"table-wrap\\">
    <Form(Search2) defaultFields={{...}} onSubmit={[Function]} onReset={[Function]} />
    <Table rowKey=\\"softwareId\\" columns={{...}} rowSelection={{...}} dataSource={{...}} pagination={{...}} useFixedHeader={false} className=\\"\\" size=\\"default\\" loading={false} bordered={false} indentSize={20} locale={{...}} showHeader={true} sortDirections={{...}} childrenColumnName=\\"children\\" />
    <Form className=\\"filter-form new-flex-layout\\" layout=\\"inline\\" onSubmit={[Function: onSubmit]} colon={true} hideRequiredMark={false}>
      <FormItem labelCol={{...}} wrapperCol={{...}} label=\\"下一步骤\\" hasFeedback={false}>
        <span value={[undefined]} id=\\"666\\" onChange={[Function: bound bound onCollect]} data-__meta={{...}} data-__field={{...}}>
          基准配置
        </span>
      </FormItem>
      <FormItem labelCol={{...}} wrapperCol={{...}} label=\\"下一步执行人\\" hasFeedback={false}>
        <Select showSearch={true} allowClear={true} getPopupContainer={[Function: getPopupContainer]} placeholder=\\"请选择\\" optionFilterProp=\\"children\\" transitionName=\\"slide-up\\" choiceTransitionName=\\"zoom\\" value={[undefined]} id=\\"baselineConfigUserId\\" onChange={[Function: bound bound onCollectValidate]} data-__meta={{...}} data-__field={{...}}>
          <Option value=\\"all\\">
            全部
          </Option>
          <Option value=\\"1\\">
            duuliy
          </Option>
        </Select>
      </FormItem>
      <FormItem labelCol={{...}} wrapperCol={{...}} label=\\"备注\\" hasFeedback={false}>
        <Input autoComplete=\\"off\\" className=\\"filter-form-item\\" placeholder=\\"请输入\\" type=\\"text\\" disabled={false} value={[undefined]} id=\\"memo\\" onChange={[Function: bound bound onCollect]} data-__meta={{...}} data-__field={{...}} />
      </FormItem>
    </Form>
  </div>
</Connect(Form(commonModal))>"
`;
